rust/arrow/src/ipc/file/reader.rs

```diff
@@ -525,6 +643,80 @@ mod tests {
+        0x9b, 0x99, 0x38, 0x66, 0xfd, 0x41, 0xd8, 0x8b, 0x49, 0x7d, 0x6d, 0xb5, 0x71,
+        0x77, 0xb5, 0xe9, 0x8c, 0x29, 0x6a, 0x31, 0xb4, 0xa1, 0x85, 0x41, 0x7a, 0x8e,
+        0xed, 0x2b, 0xac, 0x1d, 0x10, 0x71, 0xfb, 0x75, 0x32, 0x99, 0x60, 0xf4, 0x16,
+        0x6e, 0xc9, 0xf8, 0x35, 0x65, 0xf4, 0x7f, 0xbb, 0x20, 0x1c, 0x8c, 0x36, 0x7e,
+        0xae, 0x63, 0x9d, 0x61, 0xcd, 0x63, 0x7b, 0x19, 0x63, 0xbc, 0x0c, 0x4b, 0x7c,
+        0x66, 0x55, 0xe6, 0x07, 0xa3, 0x50, 0xd3, 0x6c, 0x5d, 0xa0, 0xad, 0xb6, 0x1c,
+        0xd3, 0x36, 0x13, 0x16, 0x08, 0xbf, 0xa5, 0x26, 0xf6, 0xeb, 0x32, 0x57, 0xd8,
+        0x58, 0xcf, 0x10, 0xa7, 0xb9, 0x05, 0x39, 0xb7, 0x33, 0x11, 0x3c, 0x2c, 0x39,
+        0xfe, 0x84, 0x63, 0xb2, 0x43, 0xf7, 0x87, 0xc6, 0x1f, 0xa5, 0x1d, 0xb2, 0xdf,
+        0x92, 0xf9, 0x66, 0x85, 0xdb, 0xab, 0x62, 0x35, 0xb9, 0xdd, 0x8a, 0xcc, 0x4f,
+        0x54, 0xde, 0x69, 0xf6, 0x4c, 0xfe, 0x27, 0xb8, 0x51, 0x90, 0x00, 0xb7, 0x51,
+        0x4b, 0xd7, 0x45, 0x6b, 0x4c, 0xd6, 0xf5, 0x08, 0xbf, 0xf7, 0xa7, 0x7c, 0x58,
+        0x32, 0x13, 0x66, 0xb9, 0xcf, 0xc2, 0x59, 0xd2, 0x76, 0xea, 0x83, 0x32, 0x20,
+        0x71, 0xa1, 0x3c, 0xa8, 0xec, 0xc3, 0xba, 0x14, 0x7e, 0x86, 0x2c, 0xed, 0x56,
+        0xb0, 0x27, 0x92, 0x6e, 0x85, 0x9b, 0x84, 0x42, 0x91, 0xcb, 0x6a, 0x52, 0x6e,
+        0xc1, 0x8d, 0x74, 0x59, 0x35, 0x65, 0x9b, 0x28, 0x2d, 0xd7, 0x64, 0xea, 0xa5,
+        0xb4, 0xbc, 0x25, 0x69, 0x66, 0x72, 0x80, 0x22, 0xe4, 0x97, 0xaa, 0xd6, 0x5a,
+        0x8a, 0xcc, 0xd0, 0xec, 0xa8, 0xa5, 0x00, 0xd9, 0x31, 0x7b, 0x9e, 0xef, 0xff,
+        0xeb, 0x7a, 0x8c, 0xd4, 0xf2, 0xe5, 0xac, 0x47, 0x95, 0x79, 0xfb, 0xa2, 0x7d,
+        0xf3, 0x17, 0xeb, 0xc9, 0xfc, 0x10, 0x5b, 0x33, 0xfd, 0xe9, 0xd3, 0xd7, 0x91,
+        0x5c, 0x1f, 0xd5, 0x7f, 0x1e, 0x17, 0x95, 0xff, 0xef, 0x78, 0x27, 0x7b, 0x96,
+        0xc8, 0xb6, 0x0c, 0x95, 0xef, 0xcd, 0x69, 0xbe, 0xc5, 0x22, 0xdf, 0xe7, 0xbd,
+        0x74, 0xf6, 0x34, 0xbf, 0x2f, 0xd2, 0xa7, 0x09, 0xa0, 0x68, 0xe4, 0x17, 0x4b,
+        0x6b, 0xf3, 0x70, 0xfd, 0x1c, 0x6a, 0xda, 0xd8, 0xd6, 0xe6, 0xdb, 0x5a, 0xdb,
+        0xd6, 0xe4, 0x6d, 0x4d, 0x4e, 0x85, 0x52, 0x9c, 0xa9, 0xe9, 0xe5, 0x8d, 0x8d,
+        0x3f, 0xc8, 0xb5, 0x62, 0xa8, 0x93, 0xae, 0x26, 0xe8, 0xc4, 0xe0, 0xc8, 0xcf,
+        0xef, 0x1f, 0x5e, 0x1f, 0x3e, 0xb8, 0x6d, 0x4c, 0x64, 0xc9, 0xb3, 0xa5, 0x9f,
+        0x53, 0x72, 0x7f, 0x69, 0x6f, 0xc4, 0xbd, 0xb2, 0x78, 0xc2, 0xa9, 0x48, 0x29,
+        0xa5, 0xe9, 0x15, 0xc8, 0x9e, 0x57, 0xe7, 0x4b, 0x42, 0x7f, 0x5b, 0x1a, 0xda,
+        0x53, 0x6c, 0x93, 0xad, 0x97, 0x25, 0xd1, 0x3f, 0x93, 0x58, 0xb2, 0xc7, 0xdb,
+        0xe5, 0xf9, 0xb8, 0x2d, 0xed, 0xbd, 0x55, 0x78, 0x88, 0xb2, 0x78, 0xe2, 0x2d,
+        0x9e, 0x78, 0x8b, 0x27, 0xde, 0xe2, 0x89, 0xb7, 0x78, 0xe2, 0xfd, 0x8f, 0x27,
+        0xde, 0x0f, 0x3c, 0x98, 0x4d, 0xfe, 0x3f, 0xf6, 0x2f, 0x7a, 0x3f, 0x1b, 0x5b,
+        0x6a, 0x0f, 0x00, 0x00,
+    ];
+
     #[test]
     fn read_generated_files() {
```

Comment:
Hi @andy-thomason, my review will be a bit slow, but I'll aim to complete it by Saturday morning GMT.

I ran your branch locally, and as a litmus test I tried adding the `generated_dictionary` files from the integration testing (see https://github.com/apache/arrow/pull/5926/files#diff-cd3519a5e748548b5323b0d05d8e9c8aR724) to the `paths = vec![]` list. The test fails to successfully read the dictionary file, and seems the error is with the schema.

For some context, there is integration test data in https://github.com/apache/arrow-testing (data/arrow-ipc/integration/0.14.1). As we support more data types (and are able to read more file types) we add them to the `paths = vec![]` list.

Would you be able to add `generated_dictionary` to the list of files being tested, and look at the details of the test failure? You might not even need the `TEST_DICTIONARY` and its test.

At a high level it might be related to how we're parsing the Dictionary schema from JSON in this PR.

Thanks :)

Reply:
Hi @nevi-me, I could not figure out how to get the generated files to work. Is there some documentation? Where are these files or how do you generate them. Is there a shell script to run this test?

Reply:
Found them... https://github.com/apache/arrow-testing/tree/master/data/arrow-ipc/integration/0.14.1  I'll try to make the generated file test run now...

Reply:
You'll have to set the following environment variables so that the files can be located correctly on unit tests:
`PARQUET_TEST_DATA` and `ARROW_TEST_DATA`

Reply:
The problem seems to be that the Field struct needs to contain a dictionary field which serialises to JSON. We can change the way it is implemented, but the Field struct is used everywhere :(

Reply:
If you create a separate branch on you fork, I can collaborate with you on the work. I'd rather we make the required changes sooner so we don't have a lot of breaking changes later.

I've been considering a change too to the `DataType::Timestamp` type, because it can carry an optional timezone, but creating a separate `TimestampTZ` type seems repetitive. I haven't had a chance to put a draft PR together for comment.